From fbded7ec91a0747b4d9c9059ef48aebb5d5fc4e8 Mon Sep 17 00:00:00 2001
From: Joachim Wiberg <troglobit@gmail.com>
Date: Tue, 27 Aug 2024 13:32:18 +0200
Subject: [PATCH 16/17] Fix #89: data corruption using .conf file listen
 directive
Organization: Addiva Elektronik

The addpeer() function must save a copy of the name:service argument,
otherwise, when called from cfparse(), the contents of the string will
be corrupted.

Also, add a delpeer() function to wrap the logic for when closing and
deleting peers when stopping or reloading syslogd.

Signed-off-by: Joachim Wiberg <troglobit@gmail.com>
---
 src/syslogd.c | 47 +++++++++++++++++++++++++++++++++--------------
 src/syslogd.h |  4 ++--
 2 files changed, 35 insertions(+), 16 deletions(-)

diff --git a/src/syslogd.c b/src/syslogd.c
index fecf281..9b8dc24 100644
--- a/src/syslogd.c
+++ b/src/syslogd.c
@@ -298,12 +298,43 @@ static int addpeer(struct peer *pe0)
 	pe = calloc(1, sizeof(*pe));
 	if (pe == NULL)
 		err(1, "malloc failed");
+
 	*pe = *pe0;
+	if (pe0->pe_name)
+		pe->pe_name = strdup(pe0->pe_name);
+	if (pe0->pe_serv)
+		pe->pe_serv = strdup(pe0->pe_serv);
+
 	TAILQ_INSERT_TAIL(&pqueue, pe, pe_link);
 
 	return 0;
 }
 
+static void close_socket(struct peer *pe)
+{
+	for (size_t i = 0; i < pe->pe_socknum; i++) {
+		if (pe->pe_mode & 01000)
+			NOTE("Closing inet socket %s:%s", pe->pe_name ?: "*", pe->pe_serv);
+		socket_close(pe->pe_sock[i]);
+	}
+	pe->pe_socknum = 0;
+}
+
+static void delpeer(struct peer *pe)
+{
+	if (!pe)
+		return;
+
+	close_socket(pe);
+
+	if (pe->pe_name)
+		free(pe->pe_name);
+	if (pe->pe_serv)
+		free(pe->pe_serv);
+
+	free(pe);
+}
+
 static void sys_seqno_load(void)
 {
 	char buf[32], *str;
@@ -938,16 +969,6 @@ static int create_inet_socket(struct peer *pe)
 	return 0;
 }
 
-static void close_socket(struct peer *pe)
-{
-	for (size_t i = 0; i < pe->pe_socknum; i++) {
-		if (pe->pe_mode & 01000)
-			NOTE("Closing inet socket %s:%s", pe->pe_name ?: "*", pe->pe_serv);
-		socket_close(pe->pe_sock[i]);
-	}
-	pe->pe_socknum = 0;
-}
-
 void untty(void)
 {
 #ifdef HAVE_SETSID
@@ -2589,8 +2610,7 @@ void die(int signo)
 	 */
 	TAILQ_FOREACH_SAFE(pe, &pqueue, pe_link, next) {
 		TAILQ_REMOVE(&pqueue, pe, pe_link);
-		close_socket(pe);
-		free(pe);
+		delpeer(pe);
 	}
 
 	/*
@@ -2907,8 +2927,7 @@ static void init(void)
 			continue;
 
 		TAILQ_REMOVE(&pqueue, pe, pe_link);
-		close_socket(pe);
-		free(pe);
+		delpeer(pe);
 	}
 
 	Initialized = 1;
diff --git a/src/syslogd.h b/src/syslogd.h
index d74c0ed..0dce0ed 100644
--- a/src/syslogd.h
+++ b/src/syslogd.h
@@ -226,8 +226,8 @@
  */
 struct peer {
 	TAILQ_ENTRY(peer)	pe_link;
-	const char	*pe_name;
-	const char	*pe_serv;
+	char		*pe_name;
+	char		*pe_serv;
 	int		 pe_mark;
 	mode_t		 pe_mode;
 	int		 pe_sock[16];
-- 
2.43.0

